### PR TITLE
fix(nn): fix debug_infs error in dot_product_attention with float64

### PR DESCRIPTION
**Pull request by Moriyuki-S**
## Description

Fixes #34107 

This PR addresses an issue where `dot_product_attention` raises a `FloatingPointError` when using `float64` inputs with `jax.debug_infs(True)`.
The error occurred because the mask values generated based on `float64` limits overflowed when cast to `float32`.

## Changes

- Updated `_apply_masks` to accept an optional `mask_dtype` argument.
- Modified `_dot_product_attention_core` to pass `mask_dtype=np.float32`.
- Added a regression test case in `tests/nn_test.py`.

## Verification

- Verified the fix locally with the reproduction script.
- Added `testDotProductAttentionDebugInfsFloat64` to `tests/nn_test.py`.

### CODE REVIEW

File: tests/nn_test.py [hunk 1]
```diff
@@ -380,6 +380,17 @@ def bwd_ans(x, bias, mask):
...
+    with jax.enable_x64(True), jax.debug_infs(True):
+      dtype = jnp.float64
+      shape = (1, 2, 4)
+      q = jnp.zeros(shape, dtype=dtype)
+      k = jnp.zeros(shape, dtype=dtype)
+      v = jnp.zeros(shape, dtype=dtype)
+      mask = jnp.array([[True, False]])
+
+      nn.dot_product_attention(q, k, v, mask=mask)
+
```
Comment (burnpanck) [inline tests/nn_test.py]: If possible, parametrise this test to run for each of  `jnp.float16`, `jnp.float32` and `jnp.float64` - the 16-bit test will be broken with this current fix.

File: jax/_src/nn/functions.py [hunk 2]
```diff
@@ -902,7 +902,7 @@ def _apply_masks(logits, mask, is_causal, q_seqlen, kv_seqlen,
     mask = _get_padding_mask_logits(T, S, q_seqlen, kv_seqlen)
     combined_mask = jnp.logical_and(combined_mask, mask)
 
-  large_negative_number = _get_large_negative(logits.dtype)
+  large_negative_number = _get_large_negative(mask_dtype or logits.dtype).astype(logits.dtype)
```
Comment (burnpanck) [inline jax/_src/nn/functions.py]: This now breaks in the same way for 16 bit floats. Ultimately, the original failing case was:
`_get_large_negative(jnp.float64).astype(jnp.float32)`, now, the failing case is `_get_large_negative(jnp.float32).astype(jnp.float16).astype(jnp.float32)` (in both cases, the last `astype` is in fact in the parent function below). 

A correct solution close to this one would be to rename `mask_dtype` to something like `limit_dtype` and then dynamcially choose either `limit_dtype` or `logits.dtype`, whichever has the lower maxmimum representable value. However, there are two other feasible approaches.

File: jax/_src/nn/functions.py [hunk 3]
```diff
@@ -948,7 +948,7 @@ def _dot_product_attention_core(query, key, value, bias, mask, is_causal,
     logits = (logits + bias).astype(logits.dtype)
 
   padded_logits = _apply_masks(logits, mask, is_causal, q_seqlen, kv_seqlen,
-                               local_window_size)
+                              local_window_size, mask_dtype=np.float32)
 
   # Softmax and it is always carried out in fp32.
   padded_logits = padded_logits.astype(np.float32)
```
Comment (burnpanck) [inline jax/_src/nn/functions.py:957]: Instead of changing `_apply_masks` so that it can "anticipate" the subsequent cast to `float32`, it would probably be better to pull the cast to `float32` before the call to `_apply_masks` *only when the original logits dtype is wider than 32 bits*.

**Review by burnpanck — COMMENTED** [review]
In the current form, this PR changes behaviour for narrow dtypes. It can be fixed so that it preserves existing behaviour while fixing 64-bit behaviour exactly as in this PR.

But perhaps a discussion is warranted as to if the solution shouldn't really be not to cast the `padded_logits` to 32 bit float at all when the input is wider than that. After all, the reason of the cast to 32 bit is in order to prevent precision loss when using narrower dtypes. Extending this argument to 64 bit would call for removal of that cast in the 64 bit case. However, that would again constitute a change of behaviour for the non-masked cases, so that needs maintainer input.

File: jax/_src/nn/functions.py (hunk 2 above)
Comment (Moriyuki-S) [reply jax/_src/nn/functions.py]: As you mentioned, I added the parameterized tests for `jnp.float16`, `jnp.float32`, and `jnp.float64`.
The tests passed even for `float16` with the current implementation. This is because `_dot_product_attention_core` internally promotes types to `float32` at the very beginning:
```python
logits_dtype = jnp.promote_types(query.dtype, np.float32)
```

So `logits` are effectively `float32` by the time the mask is applied.

File: jax/_src/nn/functions.py (hunk 3 above)
Comment (Moriyuki-S) [reply jax/_src/nn/functions.py:957]: I've updated the PR to cast to `float32` before `_apply_masks` as you suggested. This drops precision for `float64` inputs, but fixes the immediate crash within current design constraints.

**Comment by Moriyuki-S** [comment]
@burnpanck 
Thank you for the insightful review. I fully agree with your points regarding the discard of precision (casting float64 to float32).
Given these architectural concerns, I would like to wait for a maintainer's input.
